contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql

```diff
@@ -3222,3 +3222,375 @@ BEGIN
+	DECLARE @tmptab VARCHAR(63) = 'sp_babelfish_autoformat' + REPLACE(NEWID(), '-', '')
+	DECLARE @tmptab2 VARCHAR(63) = 'sp_babelfish_autoformat' + REPLACE(NEWID(), '-', '')
+	DECLARE @cmd VARCHAR(1000) = 'SELECT * INTO ' + @tmptab + ' FROM ' + @tab
+	EXECUTE(@cmd)
+
+	-- Get the columns
+	SELECT 
+	   c.name AS colname, c.colid AS colid, t.name AS basetype, 0 AS maxlen
+	INTO #sp_bbf_autoformat
+	FROM sys.syscolumns c left join sys.systypes t 
+	ON c.xusertype = t.xusertype		
+	WHERE c.id = object_id(@tmptab)
+	ORDER BY c.colid
+
+	-- Get max length for each column based on the data
+	DECLARE @colname VARCHAR(63), @basetype VARCHAR(63), @maxlen int
+	DECLARE c CURSOR FOR SELECT colname, basetype, maxlen FROM #sp_bbf_autoformat ORDER BY colid
+	OPEN c
+	WHILE 1=1
+	BEGIN
+		FETCH c INTO @colname, @basetype, @maxlen
+		IF @@fetch_status <> 0 BREAK
+		SET @cmd = 'DECLARE @i INT SELECT @i=ISNULL(MAX(LEN(CAST([' + @colname + '] AS VARCHAR(500)))),4) FROM ' + @tmptab + ' UPDATE #sp_bbf_autoformat SET maxlen = @i WHERE colname = ''' + @colname + ''''
+		EXECUTE(@cmd)
+	END
+	CLOSE c
+	DEALLOCATE c
+
+	-- Generate the final SELECT
+	DECLARE @selectlist VARCHAR(8000) = ''
+	DECLARE @collist VARCHAR(8000) = ''
+	DECLARE @fmtstart VARCHAR(30) = ''
+	DECLARE @fmtend VARCHAR(30) = ''
+	OPEN c
+	WHILE 1=1
+	BEGIN
+		FETCH c INTO @colname, @basetype, @maxlen
+		IF @@fetch_status <> 0 BREAK
+		IF LEN(@colname) > @maxlen SET @maxlen = LEN(@colname)
+		IF @maxlen <= 0 SET @maxlen = 1
+		
+		IF (CHARINDEX(',' + UPPER(@colname) + ',', @hiddencols) > 0) OR (CHARINDEX(',[' + UPPER(@colname) + '],', @hiddencols) > 0) 
+		BEGIN
+			SET @selectlist += ' [' + @colname + '],'			
+		END
+		ELSE 
+		BEGIN
+			SET @fmtstart = ''
+			SET @fmtend = ''
+			IF @basetype IN ('tinyint', 'smallint', 'int', 'bigint', 'decimal', 'numeric', 'real', 'float') 
+			BEGIN
+				SET @fmtstart = 'CAST(right(space('+CAST(@maxlen AS VARCHAR)+')+'
+				SET @fmtend = ','+CAST(@maxlen AS VARCHAR)+') AS VARCHAR(' + CAST(@maxlen AS VARCHAR) + '))'
+			END
+
+			SET @selectlist += ' '+@fmtstart+'CAST([' + @colname + '] AS VARCHAR(' + CAST(@maxlen AS VARCHAR) + '))'+@fmtend+' AS [' + @colname + '],'
+			SET @collist += '['+@colname + '],'
+		END
+	END
+	CLOSE c
+	DEALLOCATE c
+
+	-- Remove redundant commas
+	SET @collist = SUBSTRING(@collist, 1, LEN(@collist)-1)
+	SET @selectlist = SUBSTRING(@selectlist, 1, LEN(@selectlist)-1)	
+	SET @selectlist = 'SELECT ' + @selectlist + ' INTO ' + @tmptab2 + ' FROM ' + @tmptab + ' ' + @orderby
+	EXECUTE(@selectlist)
+	EXECUTE('SELECT ' + @collist + ' FROM ' + @tmptab2)
+	
+	-- PRINT rowcount
+	SET @rc = @@rowcount
+	IF @printrc = 1
+	BEGIN
+		PRINT '   '
+		SET @cmd = '(' + CAST(@rc AS VARCHAR) + ' rows affected)'
+		PRINT @cmd
+	END
+	
+	-- Cleanup
+	EXECUTE('DROP TABLE ' + @tmptab)
+	EXECUTE('DROP TABLE ' + @tmptab2)
+	RETURN
+END
+$$;
+GRANT EXECUTE ON PROCEDURE sys.sp_babelfish_autoformat(IN VARCHAR(257), IN VARCHAR(1000), sys.bit, VARCHAR(1000)) TO PUBLIC;
```

Comment:
is this intended? sp_babelfish_autoformat being accessible to everyone?

Reply:
Yes, this is intentional. It's a convenient way to display the contents of any table or view in sqlcmd (try 'sysobjects'). The added value of this proc is just the formatting, it does not provide access to data that would otherwise not be accessible, so any user can benefit from this.